File: hindclaw-extension/hindclaw_ext/template_ref.py
"""Template reference parsing — reusable by HTTP endpoints and CLI.

A template reference is a string like "server/hindclaw/backend-python" that
identifies a template by scope, optional source, and name.

Format:
    scope/source/name   — sourced template (installed from marketplace)
    scope/name          — custom template (created locally, no source)

Valid scopes: "server", "personal".
"""

from pydantic import BaseModel

_VALID_SCOPES = frozenset({"server", "personal"})


class TemplateRef(BaseModel):
    """Parsed template reference.

    Attributes:
        scope: Template visibility scope ("server" or "personal").
        source: Marketplace source name (None for custom templates).
        name: Template name within the scope/source.
    """

    scope: str
    source: str | None
    name: str

    def __str__(self) -> str:
        """Format as a template reference string.

        Returns:
            "scope/source/name" for sourced, "scope/name" for custom.
        """
        if self.source:
            return f"{self.scope}/{self.source}/{self.name}"
        return f"{self.scope}/{self.name}"


def parse_template_ref(ref: str) -> TemplateRef:
    """Parse a template reference string into its components.

    Args:
        ref: Template reference (e.g., "server/hindclaw/backend-python").

    Returns:
        Parsed TemplateRef with scope, source, and name.

    Raises:
        ValueError: If the reference format is invalid.
    """
    if not ref:
        raise ValueError("Template reference must not be empty")

    parts = ref.split("/")

    if len(parts) < 2:
        raise ValueError(
            f"Invalid template reference {ref!r}: must include scope (e.g., 'server/name' or 'server/source/name')"
        )

    if len(parts) > 3:
        raise ValueError(f"Invalid template reference format {ref!r}: expected 'scope/name' or 'scope/source/name'")

    scope = parts[0]
    if scope not in _VALID_SCOPES:
        raise ValueError(f"Invalid scope {scope!r}: must be one of {sorted(_VALID_SCOPES)}")

    if len(parts) == 3:
        source = parts[1]
        name = parts[2]
        if not source:
            raise ValueError(f"Invalid template reference {ref!r}: source must not be empty")
    else:
        source = None
        name = parts[1]

    if not name:
        raise ValueError(f"Invalid template reference {ref!r}: name must not be empty")

    return TemplateRef(scope=scope, source=source, name=name)

File: packages/cli/src/template_ref.rs
use anyhow::{bail, Result};

/// Parsed template reference from CLI argument.
///
/// Formats:
///   `scope/name`                — custom template (no source)
///   `scope/source/name`         — marketplace template
#[derive(Debug, Clone)]
pub struct TemplateRef {
    pub scope: String,
    pub source: Option<String>,
    pub name: String,
}

impl TemplateRef {
    /// Parse a template reference string.
    ///
    /// Two-part: `server/my-template` → scope=server, source=None, name=my-template
    /// Three-part: `server/hindclaw/backend-python` → scope=server, source=hindclaw, name=backend-python
    pub fn parse(s: &str) -> Result<Self> {
        let parts: Vec<&str> = s.split('/').collect();
        match parts.len() {
            2 => {
                validate_scope(parts[0])?;
                if parts[1].is_empty() {
                    bail!("Invalid template reference '{}': name cannot be empty.", s);
                }
                Ok(TemplateRef {
                    scope: parts[0].to_string(),
                    source: None,
                    name: parts[1].to_string(),
                })
            }
            3 => {
                validate_scope(parts[0])?;
                if parts[1].is_empty() {
                    bail!("Invalid template reference '{}': source cannot be empty.", s);
                }
                if parts[2].is_empty() {
                    bail!("Invalid template reference '{}': name cannot be empty.", s);
                }
                Ok(TemplateRef {
                    scope: parts[0].to_string(),
                    source: Some(parts[1].to_string()),
                    name: parts[2].to_string(),
                })
            }
            _ => bail!(
                "Invalid template reference '{}'. Expected 'scope/name' or 'scope/source/name'.",
                s
            ),
        }
    }
}

fn validate_scope(scope: &str) -> Result<()> {
    match scope {
        "server" | "personal" => Ok(()),
        _ => bail!("Invalid scope '{}'. Must be 'server' or 'personal'.", scope),
    }
}

impl std::fmt::Display for TemplateRef {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.source {
            Some(src) => write!(f, "{}/{}/{}", self.scope, src, self.name),
            None => write!(f, "{}/{}", self.scope, self.name),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_two_part() {
        let r = TemplateRef::parse("server/my-template").unwrap();
        assert_eq!(r.scope, "server");
        assert!(r.source.is_none());
        assert_eq!(r.name, "my-template");
    }

    #[test]
    fn test_parse_three_part() {
        let r = TemplateRef::parse("personal/hindclaw/backend-python").unwrap();
        assert_eq!(r.scope, "personal");
        assert_eq!(r.source.as_deref(), Some("hindclaw"));
        assert_eq!(r.name, "backend-python");
    }

    #[test]
    fn test_invalid_scope() {
        assert!(TemplateRef::parse("global/my-template").is_err());
    }

    #[test]
    fn test_single_part_fails() {
        assert!(TemplateRef::parse("my-template").is_err());
    }

    #[test]
    fn test_display_two_part() {
        let r = TemplateRef { scope: "server".into(), source: None, name: "test".into() };
        assert_eq!(r.to_string(), "server/test");
    }

    #[test]
    fn test_display_three_part() {
        let r = TemplateRef { scope: "server".into(), source: Some("hindclaw".into()), name: "test".into() };
        assert_eq!(r.to_string(), "server/hindclaw/test");
    }

    #[test]
    fn test_too_many_segments() {
        assert!(TemplateRef::parse("server/source/name/extra").is_err());
    }

    #[test]
    fn test_empty_source() {
        assert!(TemplateRef::parse("server//name").is_err());
    }

    #[test]
    fn test_empty_name() {
        assert!(TemplateRef::parse("server/").is_err());
        assert!(TemplateRef::parse("server/source/").is_err());
    }
}
